src/hydra-queue-runner/hydra-queue-runner.cc

```diff
@@ -449,74 +463,131 @@ bool State::checkCachedFailure(Step::ptr step, Connection & conn)
-            }
+        to.create();
+        from.create();
 
-            MaintainCount<counter> mc(nrNotificationsInProgress);
-
-            printMsg(lvlChatty, format("sending notification about build %1%") % item.id);
-
-            auto now1 = std::chrono::steady_clock::now();
-
-            Pid pid = startProcess([&]() {
-                Strings argv;
-                switch (item.type) {
-                    case NotificationItem::Type::BuildStarted:
-                        argv = {"hydra-notify", "build-started", std::to_string(item.id)};
-                        for (auto id : item.dependentIds)
-                            argv.push_back(std::to_string(id));
-                        break;
-                    case NotificationItem::Type::BuildFinished:
-                        argv = {"hydra-notify", "build-finished", std::to_string(item.id)};
-                        for (auto id : item.dependentIds)
-                            argv.push_back(std::to_string(id));
-                        break;
-                    case NotificationItem::Type::StepFinished:
-                        argv = {"hydra-notify", "step-finished", std::to_string(item.id), std::to_string(item.stepNr), item.logPath};
-                        break;
-                };
-                execvp("hydra-notify", (char * *) stringsToCharPtrs(argv).data()); // FIXME: remove cast
-                throw SysError("cannot start hydra-notify");
-            });
-
-            int res = pid.wait();
-
-            if (!statusOk(res))
-                throw Error("notification about build %d failed: %s", item.id, statusToString(res));
-
-            auto now2 = std::chrono::steady_clock::now();
-
-            if (item.type == NotificationItem::Type::BuildFinished) {
-                auto conn(dbPool.get());
-                pqxx::work txn(*conn);
-                txn.parameterized
-                    ("update Builds set notificationPendingSince = null where id = $1")
-                    (item.id)
-                    .exec();
-                txn.commit();
+        FILE* toNotify = fdopen(dup(to.writeSide.get()), "w");
+        FILE* fromNotify = fdopen(dup(from.readSide.get()), "r");
+
+        new_pid = startProcess([&]() {
+
+            if (dup2(to.readSide.get(), STDIN_FILENO) == -1) {
+              throw SysError("cannot dup input pipe to stdin");
+            }
+            if (dup2(from.writeSide.get(), STDOUT_FILENO) == -1) {
+              throw SysError("cannot dup output pipe to stdout");
             }
 
-            nrNotificationTimeMs += std::chrono::duration_cast<std::chrono::milliseconds>(now2 - now1).count();
-            nrNotificationsDone++;
+            execlp("hydra-notify", "hydra-notify", NULL);
 
-        } catch (std::exception & e) {
-            nrNotificationsFailed++;
-            printMsg(lvlError, format("notification sender: %1%") % e.what());
-            sleep(5);
-        }
+            throw SysError("cannot start hydra-notify");
```

Comment:
What does the code that catches this do? Will it break anything that it happens in a subprocess? Will any static destructors cause problems if done in a subprocess? If so, `err` is potentially a good function to print an error message and exit without calling clean-up code.